src/analysis/history.hpp

```diff
@@ -58,21 +61,71 @@ Real ReduceOneVar(MeshData<Real> *md, const std::string &varname, int idx = 0) {
   // We choose to apply volume weighting when using the sum reduction.
   // This assumes that for "Sum" variables, the variable is densitized, meaning
   // it already contains a factor of the measure: sqrt(det(gamma))
+  Real result = 0.0;
+  Reducer_t reducer(result);
+
   const bool volume_weighting =
-      std::is_same<Reducer_t, Kokkos::Sum<Real, HostExecSpace>>::value;
+      std::is_same<Reducer_t, Kokkos::Sum<Real, HostExecSpace>>::value ||
+      std::is_same<Reducer_t, Kokkos::Sum<Real, DevExecSpace>>::value ||
+      std::is_same<Reducer_t, Kokkos::Sum<Real, Kokkos::DefaultExecutionSpace>>::value;
 
+  parthenon::par_reduce(
+      parthenon::LoopPatternMDRange(), "Phoebus History for " + varname, DevExecSpace(),
+      0, pack.GetDim(5) - 1, kb.s, kb.e, jb.s, jb.e, ib.s, ib.e,
+      KOKKOS_LAMBDA(const int b, const int k, const int j, const int i, Real &lresult) {
+        // join is a Kokkos construct
+        // that automatically does the
+        // reduction operation locally
+        const auto &coords = pack.GetCoords(b);
+        const Real vol = volume_weighting ? coords.CellVolume(k, j, i) : 1.0;
+        reducer.join(lresult, pack(b, ivar.first + idx, k, j, i) * vol);
+      },
+      reducer);
+  return result;
+}
+
+template <typename Reducer_t>
+Real ReduceInGain(MeshData<Real> *md, const std::string &varname, int idx = 0) {
```

Comment:
I don't think `ReduceInGain` needs to be templated on reducer type, since we're always going to be summing. It probably also doesn't need to take in `varname` or an `index`. I would make it just take `MeshData`. If it's not templated, you can also move the implementation to `history.cpp` instead of `hpp` and remove some of the includes above.

Reply:
Why should not it take varname or index? Should not it be a general function that integrates over the gain region whatever variable we pass? sometimes we'll integrate scalar variable sometimes vector no?

Reply:
Ah I see--there are multiple things you may want to reduce over the gain region? Then yes that's fine. You can keep the string machinery + index machinery as is.

Reply:
Yes correct, okay.